### ISSUE TITLE
feature suggestion: fieldref value serialization/deserialization

### ISSUE THREAD

**Issue by CanuckNav74** [Microsoft/AL#3933]
For primitive field types (e.g. datetime, integer/option, decimal, text, code, etc) its fairly straightforward to (de)serialize a fieldref value to/from XML or JSON). For complex field types (e.g. media, mediaset, recordid) there is no easy way to accomplish this.

What if Microsoft added some new global external procedures to the fieldref object to facilitate this? This would be helpful for primitive (de)serialization as well - e.g. let NAV handle the text formatting e.g. removing thousands separators from numerics, date/time/datetime formatted as ISO8601, blobs as base64

e.g. possible suggestion for new global funcs for both XML and JSON handling. Pseudocode for function headers below

```
procedure DeserializeFromXmlElement(var pxmlElement: XmlElement; pblnValidate: Boolean)
procedure SerializeToXmlElement(): XmlElement
procedure DeserializeFromJsonObject(var pjsonObj: JsonObject; pblnValidate: Boolean)
procedure SerializeToJsonObject(): JsonObject
```

this could really open up data transfer capabilities to/from business central.

thoughts?

**Comment by rdebath** [comment]
To keep in line with the current setup this would presumably be `FORMAT(...9)` and `EVALUATE(..., 9)`
I haven't tried those ... I assume that they don't work?  Yet?

BLOBs do have some code on TempBLOB in the most recent version, but to do that for the media "type" you have to rummage around in the `Media` table. (Rather poor design IMO)

**Comment by CanuckNav74** [comment]
yeah that's what I'm doing now via the FORMAT(...) from C/AL (haven't tried in AL yet - this is my first conversion project so I'm not sure what FORMAT calls work/dont-work in BC)

Reason why I'm suggesting this enhancement is because I'm having a hard time wrapping my head around converting a customized NAV2009 customer to BC OnCloud + extensions AND have some sort of way to transport all legacy company data (which includes custom fields on base tables) to the OnCloud tenant. I'm aware that there are the rapidstart configuration templates at our disposal, but I see overhead on massaging that legacy excel data to be absolutely massive. Case in point, the client NAV2009 database I'm working with has 235 tables in the 1..49999 range that have one or more records. 

I genuinely want to migrate my legacy clients to BC, but challenges like this are going to make upgrading a hard sell. 

The thought of serializing all tables/records/fields from the legacy system with the help of recordref and fieldref and newtonsoft JSON.NET is very appealing in theory (then write up a BC Codeuit to consume that file and then re-create the table data from the JSON payload), but I can see the non-primitive fields that I mentioned (e.g. recordID, etc.) being a stumbling block.

**Comment by rdebath** [comment]
The DK version has transferred some functionality to extensions. Unfortunately, the upgrade procedure, though automatic, is rather primitive. All the tables have been renumbered and the data is copied record by record from the old tables to the new ones. For anything that isn't essentially trivial this will be a huge job to run, even if the code can be generated automatically. Though I expect it'll be faster than trying to copy via text files (which probably have to have limited size too).

**Comment by StanislawStempin** [comment]
Thank you for the suggestion. We will consider it for the long term backlog

**Comment by esbenk** [comment]
Hi! Thank you for your suggestion. We are working on focusing the activity in this repository on addressing issues related to the AL development tools. 
As this issue appears to require changes across multiple areas of our product, we recommend that you post this on our Ideas forum at aka.ms/BCIdeas, or vote up the idea if its already there.
